File: examples/18-layout-helpers/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"

	log "github.com/djinn-soul/gopptx/pkg/stdlog"

	"github.com/djinn-soul/gopptx/pkg/pptx"
	"github.com/djinn-soul/gopptx/pkg/pptx/styling"
)

const (
	outputDir  = "examples/output"
	outputFile = "18_layout_helpers.pptx"
)

func main() {
	if err := run(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run() error {
	if err := os.MkdirAll(outputDir, 0o750); err != nil {
		return fmt.Errorf("create output dir: %w", err)
	}

	builder := pptx.NewPresentationBuilder("Layout Helpers Demo").
		WithSlideSize(pptx.SlideSize16x9())

	// Slide 1: demonstrate unit conversion values
	convSlide := buildUnitConversionSlide()
	builder.AddSlide(convSlide)

	// Slide 2: shapes positioned using styling.Inches
	posSlide := buildPositionedShapesSlide()
	builder.AddSlide(posSlide)

	// Slide 3: demonstrate Emu and retrieval helpers
	emuSlide := buildEmuHelpersSlide()
	builder.AddSlide(emuSlide)

	outputPath := filepath.Join(outputDir, outputFile)
	if err := builder.WriteToFile(outputPath); err != nil {
		return fmt.Errorf("write: %w", err)
	}

	log.Printf("Generated %s\n", outputPath)
	return nil
}

// buildUnitConversionSlide shows the unit-conversion helpers and their values.
func buildUnitConversionSlide() pptx.SlideContent {
	// styling.Inches / styling.InchesToEMU / styling.Emu
	oneInch := styling.Inches(1)
	twoInch := styling.InchesToEMU(2) // alias for styling.Inches(2)
	rawEmu := styling.Emu(914400)     // 1 inch in EMU
	_ = twoInch

	slide := pptx.NewSlide("Unit Conversion Helpers").
		AddBullet(fmt.Sprintf("styling.Inches(1)     = %d EMU (%.4f in)", oneInch.Emu(), oneInch.Inches())).
		AddBullet(fmt.Sprintf("styling.InchesToEMU(2) = %d EMU (%.4f in)", twoInch.Emu(), twoInch.Inches())).
		AddBullet(fmt.Sprintf("styling.Emu(914400)   = %d EMU (%.4f in)", rawEmu.Emu(), rawEmu.Inches())).
		AddBullet(fmt.Sprintf("styling.Centimeters(2.54) = %d EMU", styling.Centimeters(2.54).Emu())).
		AddBullet(fmt.Sprintf("styling.Points(72)    = %d EMU", styling.Points(72).Emu()))

	return slide
}

// buildPositionedShapesSlide places two shapes using styling.Inches for precise coordinates.
func buildPositionedShapesSlide() pptx.SlideContent {
	// Shape at (1in, 1in) with size 2in × 1in
	x1 := styling.Inches(1)
	y1 := styling.Inches(1)
	w := styling.Inches(2)
	h := styling.Inches(1)

	// Shape at (4in, 2in) with size 2in × 1in
	x2 := styling.Inches(4)
	y2 := styling.Inches(2)

	slide := pptx.NewSlide("Precise Positioning with styling.Inches").
		AddShape(
			pptx.NewShape(pptx.ShapeTypeRectangle, x1, y1, w, h).
				WithFill(pptx.NewShapeFill("4472C4")).
				WithText(fmt.Sprintf("(%.0fin, %.0fin) 2×1in", x1.Inches(), y1.Inches())),
		).
		AddShape(
			pptx.NewShape(pptx.ShapeTypeRoundedRectangle, x2, y2, w, h).
				WithFill(pptx.NewShapeFill("ED7D31")).
				WithText(fmt.Sprintf("(%.0fin, %.0fin) 2×1in", x2.Inches(), y2.Inches())),
		)

	return slide
}

// buildEmuHelpersSlide demonstrates retrieving values back from a Length.
func buildEmuHelpersSlide() pptx.SlideContent {
	oneInch := styling.Inches(1)
	rawEmu := styling.Emu(914400)

	slide := pptx.NewSlide("Retrieving Values from Length").
		AddBullet(fmt.Sprintf("oneInch.Inches() = %.6f", oneInch.Inches())).
		AddBullet(fmt.Sprintf("oneInch.Emu()    = %d", oneInch.Emu())).
		AddBullet(fmt.Sprintf("oneInch.Cm()     = %.4f", oneInch.Cm())).
		AddBullet(fmt.Sprintf("oneInch.Pt()     = %.4f", oneInch.Pt())).
		AddBullet(fmt.Sprintf("rawEmu.Inches()  = %.6f (should equal 1.0)", rawEmu.Inches())).
		AddBullet(fmt.Sprintf("rawEmu.Emu()     = %d", rawEmu.Emu()))

	return slide
}

File: examples/18-layout-helpers/main.py
"""Demonstrate layout helpers including unit conversion and precise positioning.

This example demonstrates:
- Using Inches() for precise EMU-based coordinate calculations
- Positioning shapes with exact inch measurements
- Documenting unit conversion values (Inches, Centimeters, Points)
"""

from __future__ import annotations

from pathlib import Path

from gopptx import Presentation
from gopptx.presentation.slides import SlideLayoutType
from gopptx.schemas import Inches


def _add_unit_conversion_slide(prs: Presentation) -> None:
    """Add a slide showing unit conversion helper values."""
    prs.add_bullet_slide(
        "Unit Conversion Helpers",
        [
            f"Inches(1)          = {Inches(1)} EMU",
            f"Inches(2)          = {Inches(2)} EMU",
            "Centimeters(2.54)  = ~914400 EMU  (= 1 inch)",
            "Points(72)         = ~914400 EMU  (= 1 inch)",
            "All coordinates passed to gopptx use Inches() values.",
        ],
    )


def _add_positioned_shapes_slide(prs: Presentation) -> None:
    """Add a slide with shapes placed at precise inch coordinates."""
    prs.add_slide("Precise Positioning with Inches()", layout=SlideLayoutType.BLANK)
    idx = prs.slide_count - 1

    prs.add_shape(
        idx,
        "RECTANGLE",
        bounds=(Inches(1), Inches(1), Inches(2), Inches(1)),
        text="(1in, 1in) 2x1in",
        properties={"fill": "4472C4"},
    )
    prs.add_shape(
        idx,
        "ROUNDED_RECTANGLE",
        bounds=(Inches(4), Inches(2), Inches(2), Inches(1)),
        text="(4in, 2in) 2x1in",
        properties={"fill": "ED7D31"},
    )


def _add_emu_helpers_slide(prs: Presentation) -> None:
    """Add a slide demonstrating value retrieval from inch measurements."""
    one_inch = Inches(1)
    prs.add_bullet_slide(
        "Retrieving Values from Inches()",
        [
            f"Inches(1) value = {one_inch}",
            "Pass Inches(n) directly to x, y, w, h parameters.",
            "1 inch = 914,400 EMU (English Metric Units).",
            "Use consistent units across all shape positioning calls.",
        ],
    )


def main() -> None:
    """Create a presentation demonstrating layout helper usage."""
    output_dir = Path("examples/output")
    output_dir.mkdir(exist_ok=True)

    with Presentation.new("Layout Helpers Demo") as prs:
        _add_unit_conversion_slide(prs)
        _add_positioned_shapes_slide(prs)
        _add_emu_helpers_slide(prs)

        output_path = output_dir / "18-layout-helpers.pptx"
        prs.save(str(output_path))
        print(f"Saved: {output_path}")

    print("\n=== SUMMARY ===")
    print("Created 3 slides demonstrating layout helpers:")
    print("  - Unit conversion reference (Inches, Centimeters, Points)")
    print("  - Shapes positioned with precise Inches() coordinates")
    print("  - EMU value retrieval documentation")


if __name__ == "__main__":
    main()
